vignettes/custom-expectation.Rmd

```diff
@@ -72,33 +110,75 @@ expect_vector_length <- function(object, n) {
+Or imagine if you're checking to see if an object inherits from an S3 class. In R, there's no direct way to tell if an object is an S3 object: you can confirm that it's an object, then that it's not an S4 object. So you might organise your expectation this way:
 
 ```{r}
-mtcars |>
-  expect_type("list") |>
-  expect_s3_class("data.frame") |> 
-  expect_length(11)
+expect_s3_class <- function(object, class) {
+  act <- quasi_label(rlang::enquo(object), arg = "object")
+
+  if (!is.object(act$val)) {
+    return(fail(sprintf("%s is not an object.", act$lab)))
+  }
+
+  if (isS4(act$val)) {
+    return(fail(sprintf("%s is an S4 object, not an S3 object.", act$lab)))
+  }
+
+  if (!inherits(act$val, class)) {
+    msg <- sprintf(
+      "%s inherits from %s not %s.",
+      act$lab,
+      paste0(class(object), collapse = "/"),
+      paste0(class, collapse = "/")
+    )
+    return(fail(msg))
+  }
+
+  pass(act$val)
+}
 ```
 
-## Testing your expectations
+```{r}
+#| error: true
+x1 <- 1:10
+TestClass <- methods::setClass("Test", contains = "integer")
+x2 <- TestClass()
+x3 <- factor()
+
+expect_s3_class(x1, "integer")
+expect_s3_class(x2, "integer")
+expect_s3_class(x3, "integer")
+```
 
-testthat comes with three expectations designed specifically to test expectations: `expect_success()` and `expect_failure()`:
+## Repeated code
 
-* `expect_success()` checks that your expectation emits exactly one success and zero failures.
-* `expect_failure()` checks that your expectation emits exactly one failure and zero successes. 
-* `expect_failure_snapshot()` captures the failure message in a snapshot, making it easier to review if it's useful or not.
+As you write more expectations, you might discover repeated code that you want to extract out in to a helper. Unfortunately creating helper functions is not straightforward in testthat because every `fail()` captures the calling environment in order to give maximally useful tracebacks. Because getting this right is not critical (you'll just get a slightly suboptimal traceback in the case of failure), we don't recommend bothering. However, we document it here because it's important to get it right in testthat itself.
+
+The key challenge is that `fail()` captures a `trace_env` which should be the execution environment of the expectation. This usually works, because the default value of `trace_env` is `caller_env()`. But when you introduce a helper, you'll need to explicitly pass it along:
 
 ```{r}
-test_that("expect_length works as expected", {
-  x <- 1:10
-  expect_success(expect_length(x, 10))
-  expect_failure(expect_length(x, 11))
+expect_length_ <- function(act, n, trace_env = caller_env()) {
+  act_n <- length(act$val)
+  if (act_n != n) {
+    msg <- sprintf("%s has length %i, not length %i.", act$lab, act_n, n)
+    return(fail(msg, trace_env = trace_env))
+  }
 
-  expect_snapshot_failure(expect_length(x, 11))
-})
+  pass(act$val)
```

Comment:
maybe a note should be added that both `fail()` and `pass()` should be used in the helper, as it will be used in `return()` explicitly or otherwise.